adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -289,12 +292,26 @@ private void doUploadMessage(final AmqpContext context, final Future<MessageSend
             }
             return Future.<Void> succeededFuture();
         }).recover(t -> {
-            LOG.info("Cannot process message for Device [deviceId: {}] with Tenant [tenantId: {}]",
+            LOG.debug("Cannot process message for Device [tenantId: {}, deviceId: {}, endpoint: {}]",
+                    context.getTenantId(),
                     context.getDeviceId(),
-                    context.getTenantId());
-            if (!context.remotelySettled()) {
+                    context.getEndpoint());
+            if (!context.isRemotelySettled()) {
                 // client wants to be informed that the message cannot be processed.
-                context.released();
+
+                // t can only be a ServiceInvocationException
+                final ServiceInvocationException error = (ServiceInvocationException) t;
+                final StringBuilder errorMsg = new StringBuilder();
+
+                if (ServerErrorException.class.isInstance(t)) {
+                    errorMsg.append("Releasing message: {} - {}");
+                    context.released();
+                } else {
+                    errorMsg.append("Rejecting message : {} - {}");
+                    context.reject(AmqpError.NOT_FOUND, (ClientErrorException) error);
```

Comment:
IMHO we should try to have only a single place in the code where we map ServiceInvocationExceptions (or Throwables in general) to AmqpErrors. That place should probably be the AmqpContext's *reject* method.

Reply:
i thought we are trying to reject only in the case of a client error not a server error? so we exclude mapping server errors and just release?

the possible error codes that may be reported while uploading a message:
1. Bad Request (400) maps to `hono:bad-request` (client-error)
2. Forbidden (403) maps to `amqp:not-found` (adapter is disabled for tenant - client error)
3. Internal Error (500) maps to `amqp:internal-error` (peer did not settle message - server error)
4. Unavailable (503) maps to `amqp:precondition-failed` (no credit available)

WDYT?

Reply:
Sure, we only reject messages that caused a ClientErrorException and we release messages that caused a ServerErrorException. Still, we could have a single method in `AmqpContext` which handles a `ServiceInvocationException` and thus hides this policy and the concrete mapping to AMQP errors, right? BTW, we should probably map 403 to the standard `amqp:unauthorized-access`, don't you think? 500 and 503 do not need to be mapped because we are simply releasing messages for status codes in the 500s, right?

Reply:
> Still, we could have a single method in AmqpContext which handles a ServiceInvocationException and thus hides this policy and the concrete mapping to AMQP errors, right?

I see! If I understand correctly, we now only need one call, say `AmqpContext#handleFailure(SIE)`, and do not need explicit `reject` and `release` calls for dispositions?

> amqp:unauthorized-access -> 403

Sure. i was unable to relate the `security-settings` wording in the spec to a disabled adapter.

Reply:
that would make sense to me :-)

Reply:
ok, PR update